mlir/lib/Dialect/Rock/utility/transformMapUtils.cpp

```diff
@@ -46,15 +47,41 @@ mlir::rock::untransform(OpBuilder &b, Value transformed, ArrayAttr existing) {
     transformList.push_back(transform.getTransform());
     ret = transform.getInput();
   }
-  return {ret, b.getArrayAttr(transformList)};
+
+  bool isBig = llvm::any_of(transformList, [](Attribute a) {
+    return needs64BitIndices(cast<TransformMapAttr>(a));
+  });
+  auto bufferType = cast<ShapedType>(ret.getType());
+  int64_t elemBytes;
+  if (auto shapedTy = dyn_cast<ShapedType>(bufferType.getElementType()))
+    elemBytes =
+        (shapedTy.getNumElements() * shapedTy.getElementTypeBitWidth()) / 8;
+  else
+    elemBytes = bufferType.getElementTypeBitWidth() / 8;
+
+  if (bufferType.hasStaticShape())
+    isBig = (bufferType.getNumElements() * elemBytes) >=
+            ((int64_t)1 << (int64_t)32);
+  else
+    isBig = true;
+  return {ret, b.getArrayAttr(transformList), isBig};
 }
 
-std::tuple<Value, ArrayAttr>
+std::tuple<Value, ArrayAttr, bool>
 mlir::rock::untransform(OpBuilder &b, Value transformed,
                         ArrayRef<Attribute> existing) {
   return untransform(b, transformed, b.getArrayAttr(existing));
 }
 
+bool mlir::rock::needs64BitIndices(TransformMapAttr map) {
+  // Negative lengths are some form of dynamic index and therefore big.
+  auto isBig = [](int64_t l) -> bool {
+    return l < 0 || l > (int64_t)(std::numeric_limits<int32_t>::max());
```

Comment:
same here... we might factor this out

Reply:
Where is the duplication?

Reply:
{l > (int64_t)(std::numeric_limits<int32_t>::max() } == 
 {l >= 1 << 32} , right ?

Reply:
yea you right, uint32_t::max()
(I just dont like these magic numbers; was looking a way to get rid of them)